### DISCUSSION TITLE
Graphic Engine

### DISCUSSION THREAD

**Discussion by vbBaka** [twinbasic/twinbasic#607] [General]
Im interested to understand more of the "graphic engine" that TB will use.
both for the legacy/32bit/64bit mode but also the cross-platform.

as we all know, VB6 own graphic engine is not that great. that is why we use GDI, and when that is not enough we use DirectX, Direct2D, Cairo or OpenGL.

if we have:
- a rendering mode that allow 24/32bit
- a EndDraw Mode/Event (that can be switched on/off using the Option statement) that allow us to work similar to DirectX, where no graphic is rendered until the EndDraw (that is follow the monitor-sync) creating a double-buffering mode and also a WaitForVerticalBlankEvent.
- a way to load/save png and other file formats
- and different commands to manipulate pictures, such as rotate, colors, scaling etc.

that will mean, we have everything we need to create applications that are focused on visual, such as demo, games, simulators, etc.
even if not as "strong" as DirectX, but at least on-par with GDI32.

if theres nothing or maybe even impossible to apply. Im thinking of Cairo, as it allow cross-platform, if that could be an alternative for the graphic engine that can be integrated with TB.

**Comment by WaynePhillipsEA** [comment]
Hi @vbBaka, thanks for joining us over here.   Naturally, the initial priority is to get the standard GUI work done, with not much thought having yet gone into how we will support the more complex graphics side of things (for games etc).

Cairo is certainly a possibility, though the LGPL licence means that we can't integrate it directly into built EXEs, and thus it will require a separate DLL to be shipped alongside the executables.  Of course that's not a massive problem.

At this stage I'd be interested to hear more about peoples experiences with Cairo (+RC6), or any other graphic engines.

**Comment by vbBaka** [comment]
I think, for convenience it should be a cross-platform engine.
we have cairo and opengl of what I know. Im sure theres others, but I can't say about stability, performance or if they are free or not.

to have it integrated into the exe would be the best option. but not sure how much work it means to create.
